BUG: allow missing values in Index when calling Index.sort_values

AlexKirko opened the pull request:
- [X] closes #35584, xref #35922 
- [X] tests added 1 / passed 1
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### Problem

`Index.sort_values` breaks when `Index` includes `None`. The OP points out that a `Series` doesn't break in a similar case. In my opinion, there is a number of ways missing values can creep into an Index, and `sort_values` shoudn't be breaking.

### Details

When we do `Series.sort_values` we shunt the missing values to the end or the beginning of the Series depending on `na_position`, and then `argsort` the rest. When we currently do `sort_values` on an `Index`, we try to `argsort` the whole thing, and it expectedly breaks.

### Solution and alternatives

I propose we do the same stuff for `Index` that `Series` does, that is send the missing values to the end or the beginning. Added the `na_position` kwarg that works similar to how it works in `Series`.

Another possible solution is to explicitly forbid Indices with missing values and raise a clear error. I don't think it's advisable, since there isn't much of a practical difference between non-unique indices, which we allow, and this case.

AlexKirko committed:
BUG: attempt initial fix

AlexKirko committed:
TST: add test

AlexKirko committed:
CLN: run black

AlexKirko committed:
CLN: clean up unnecessary print

AlexKirko commented:
Couldn't find a better home for the test than `test_common.py` in indixes. Maybe there is a better place for it?

AlexKirko committed:
exempt MultiIndex from handling missing values

AlexKirko commented:
Exempted MultiIndex from this fix, as it's completely unclear where you should put missing values in this case.

AlexKirko committed:
CLN: run black

AlexKirko commented:
Hm. There are a bunch of tests that expect NaNs to be at the beginning. Will have to add `na_position` to the function signature.

AlexKirko committed:
add na_position kwarg to sort_values

AlexKirko committed:
debug na_position logic in sort_values

AlexKirko committed:
CLN: fix na_position type

AlexKirko committed:
DOC: mv to 1.2.0 since API is changed

AlexKirko committed:
DOC: make the comment more specific

AlexKirko committed:
TST: parametrize the test, add raise check

AlexKirko committed:
CLN: run black

AlexKirko commented:
I think this is ready for review. Since this PR changes the API, so I put it in 1.2.0 whatsnew.

AlexKirko committed:
CLN: fix versionadded

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4371,7 +4371,11 @@ def asof_locs(self, where, mask):
+        self,
+        return_indexer=False,
+        ascending=True,
+        key: Optional[Callable] = None,
```
can you order these the same as in series (na_position before key)

AlexKirko replied:
Done.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4425,9 +4435,28 @@ def sort_values(
+                    f"Can be only 'first' or 'last', '{na_position}' given."
+                )
+        else:
+            _as = idx.argsort()
```
can you comment that this is the MI one?

it *might* be better to actually have a helper method to compute the _sort_indexer where you define for Index and MI then this method becomes simpler.

jreback replied:
also for MI we are basically ignoreing na_position right?

AlexKirko replied:
Done.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -4425,9 +4435,28 @@ def sort_values(
+                    _as[: np.sum(good)] = _as[: np.sum(good)][::-1]
+            elif na_position == "first":
+                _as = np.concatenate([_as[bad], _as[good][idx[good].argsort()]])
+                if not ascending:
```
can you share this code with the series impl (might be better / easier to create a helper to actually do this) and put it in pandas/core/sorting.py

AlexKirko replied:
@jreback Found such a function in `sorting.py`: it's called `nargsort`. With a minor addition to insure that mergesort defaults to quicksort when sorting objects, it can be used for our purposes.

But I've uncovered a bit of a problem when trying to unify `Index` and `Series` behavior: when we set `ascending=False` and the object containes duplicates, in general, we expect different behavior from `Series.sort_index` and `Series.sort_values`. We expect `sort_values` to reverse the order of equal elements, but we expect `sort_index` to be stable and maintain it. This also creates weirdness in `DataFrame.sort_values`.

I've switched `Series.sort_values` to using `nargsort` and added an argument to `nargsort` to signify if we want to preserve order or not, but there are still edge cases in the test sutie because of this inconsistency.

It is probably best to settle on one convention and carefully alter all existing tests to fit it. I personally think that equal elements should maintain order when we sort in descending order, that is, the descending sort should also be stable. I don't know if we should do this in this PR though, or leave Series as it is and only add NaN-handling to `Index.sort_values` through nargsort. Then I can make another PR to handle Series there and alter the tests to the convention we agree upon.

What do you think?

jreback reviewed (changes requested):
looks good. opportunity to clean up some code here.

AlexKirko committed:
CLN: move na_position before key

AlexKirko committed:
DOC: add comment for MultiIndex

AlexKirko committed:
REFACT: switch to nargsort from sorting.py

AlexKirko committed:
reverse equal els in nargsort of ascending is False

AlexKirko committed:
switch Series to nargsort

AlexKirko committed:
Revert "reverse equal els in nargsort of ascending is False"

Most tests expect the order of equal elements to be preserved
when ascending is False.

pep8speaks commented:
Hello @AlexKirko! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/core/sorting.py`](https://github.com/pandas-dev/pandas/blob/66120794064537bac0df185e43172d826d2c5caf/pandas/core/sorting.py):

> [Line 315:5](https://github.com/pandas-dev/pandas/blob/66120794064537bac0df185e43172d826d2c5caf/pandas/core/sorting.py#L315): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

AlexKirko committed:
add TypeError to exception catch in nargsort

AlexKirko committed:
reverse equal els order in nargsort (again)

AlexKirko committed:
CLN: isort test_datetime.py

AlexKirko committed:
keep equal el order in nargsort (again)

AlexKirko committed:
add equals_reversible to nargsort and a debugging branch

AlexKirko committed:
CLN: remove debug code

AlexKirko committed:
CLN: remove extra newline in test_datetime

AlexKirko committed:
CLN: alter test_ops to preserve equal element order

Preserve equal element order in indexes.period.test_ops.test_order_compat

AlexKirko committed:
CLN: run black

AlexKirko commented:
For some weird reason, we expect different behavior from `Index.sort_values` and `Series.sort_values` when `ascending=False` and the object contains duplicates. I'll give more details once I figure out a candidate solution that passes the tests without altering the expectation of this behavior.

AlexKirko commented:
Guess I'll limit this PR to implementing missing values support for `Index.sort_values` using `nargsort`.

Switching the Series.sort_values to it, I think, belongs in another issue and PR, along with picking a convention for sorting duplicates and enforcing it across our test suite.

AlexKirko committed:
revert changes to Series and connected tests

AlexKirko committed:
reimplement test fix for Index.sort_values

AlexKirko committed:
TST: don't expect same indexer for Index and PeriodIndex

The new sort_values implementation for Index is stable when sorting in
descending order. However, we expect PeriodIndex.sort_values to be
unstable (and have tests for this). So until we enforce the same
convention for sorting duplicate values in descending order, we
can't check that the indexers for these classes are the same.

jreback reviewed (changes requested):
see my comments on the tests; we need to expand to test all index types here (of course can xfail ones which don't work for now)

AlexKirko commented:
Great, I'll make the changes tomorrow.

jreback commented on pandas/tests/indexes/period/test_ops.py:
```diff
@@ -175,9 +175,9 @@ def _check_freq(index, expected_index):
-            exp = np.array([2, 1, 3, 4, 0])
-            tm.assert_numpy_array_equal(indexer, exp, check_dtype=False)
-            _check_freq(ordered, idx)
+            # GH 35584. Index sort is now stable when sorting in descending order
```
can you create a new test and xfail it

AlexKirko replied:
Added the new test to the end of the file.

jreback replied:
ok you don't need this comment then

AlexKirko replied:
Removed the comment.

AlexKirko committed:
TST: add Index types, switch to fixture

AlexKirko committed:
CLN: fix import order

AlexKirko committed:
TST: add test for Index and PeriodIndex incompatibility

AlexKirko commented on pandas/tests/indexes/test_common.py:
```diff
@@ -395,3 +395,26 @@ def test_astype_preserves_name(self, index, dtype, copy):
             assert result.name == index.name
+
+
+@pytest.mark.parametrize("idx", [Index(["a", None, "c", None, "e"])])
```
Made a new fixture with all kind of indices.

AlexKirko commented on pandas/tests/indexes/test_common.py:
```diff
@@ -395,3 +395,26 @@ def test_astype_preserves_name(self, index, dtype, copy):
+    # GH 35584. Test that sort_values works with missing values,
+    # sort non-missing and places missing according to na_position
+
+    if na_position not in ["first", "last"]:
```
Done.

AlexKirko commented:
@jreback
Made the chagnes to tests, all green. Please take a look.

AlexKirko commented:
@jreback 
Added a fixture, reworked the tests. All green, please take a look whether this is what you had in mind.

jreback commented on pandas/tests/indexes/period/test_ops.py:
```diff
@@ -333,3 +334,16 @@ def test_freq_setter_deprecated(self):
             idx.freq = pd.offsets.Day()
+
+
+@pytest.mark.xfail(reason="PeriodIndex.sort_values currently unstable")
```
well its doesn't work :->, do we have an issue tracking this, pls add the refernce in the xfail (and if we don't pls create an issue)

AlexKirko replied:
Couldn't find an existing issue, so I opened #35922 35922

jreback replied:
kk add the issue number to the xfail message

jreback replied:
though I think the comprehensive test below covers this? (not averse to having this as well, just asking)

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +400,61 @@ def test_astype_preserves_name(self, index, dtype):
+        tm.makeRangeIndex(10),
+        tm.makeCategoricalIndex(10),
+        tm.makeMultiIndex(10),
+        tm.makeDateIndex(10),
```
we *already* have the 'index' fixtures, pls use it.

jreback replied:
if you need a derivative one, then you can create it similary (e.g. use indices_dict). you *could* just add it in the pandas/conftest.py is fine.

AlexKirko replied:
Sorry, didn't know that we keep all the general fixtures in `pandas/conftest.py`. I'll add another fixture there that adds missing values to the indices.

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +400,61 @@ def test_astype_preserves_name(self, index, dtype):
+            index_with_missing.sort_values(na_position=na_position)
+
+
+@pytest.mark.parametrize("na_position", ["first", "last"])
```
is `na_position=None` valid? (it *should* just default or error if we don't accept it), so test either here or above if its supposed to error.

AlexKirko replied:
`na_position=None` isn't valid. Added a test case.

AlexKirko committed:
TST: move fixture to pandas/conftest.py

AlexKirko committed:
TST: add na_position=None test case

AlexKirko committed:
TST: move index type xfails to test_common.py

AlexKirko committed:
DOC: remove unnecessary comment in test_ops

AlexKirko committed:
DOC: add xref to unstable sort issue

AlexKirko commented:
Numpy dev fails now, but I don't see Index.sort_values being called in the failing tests. Will investigate why this is happening.

AlexKirko committed:
TST: add prints to debug numpy dev test fails

AlexKirko committed:
Revert "TST: add prints to debug numpy dev test fails"

This reverts commit c7e824049d82df9370126f3c6bf5737f3c8c194a.

AlexKirko commented:
So far unable to reproduce the test failures on my local machine (including using a numpy-dev environment on windows).

AlexKirko committed:
restart tests

AlexKirko committed:
Merge branch 'master' into ind-sort-values

AlexKirko committed:
DOC: remove junk from whatsnew

AlexKirko committed:
TST: warp test for debug

AlexKirko committed:
Revert "TST: warp test for debug"

This reverts commit d8f9061d480d33b485882d5c3b6e35d032411301.

AlexKirko committed:
remove new tests for debug

AlexKirko commented:
This is super-weird, but I figured out why the np-dev pipeline is failing. For some reason `test_union_different_types` in `test_setops.py' pulls my new fixture instead of `index` and `index_fixture2`. This, as far as I can tell, is a bug that's specific to this pipeline (can't reproduce it locally on Ubuntu 18.04 and numpy-dev environment).

Now that I know what's going on, I should be able to circumvent it.

AlexKirko committed:
Revert "remove new tests for debug"

This reverts commit 29d684907ecc654628f3ae2447db825a7aaaa5fa.

AlexKirko committed:
TST: try deep copy in fixture

AlexKirko committed:
TST: try immediate return instead of deep copy

AlexKirko committed:
Revert "TST: try immediate return instead of deep copy"

This reverts commit a3310ba9172db9511e7d5954c4bf498136dd8329.

AlexKirko committed:
REFACT: add immediate returns

AlexKirko committed:
Merge branch 'master' into ind-sort-values

AlexKirko committed:
DOC: add comment to conftest clarifying deep copy

AlexKirko committed:
restart tests

AlexKirko commented:
Waiting for the mypy check. The `py37_np16` appears to be unrelated too, will look into it.

AlexKirko committed:
Merge branch 'master' into ind-sort-values

AlexKirko commented:
@jreback 
Finally all green, please take another look.
Apologies for the delay: ran into a bug with `Index.copy` on one of the pipelines, which I wasn't able to reproduce. Took a while to track down.

AlexKirko committed:
CLN: remove test-output.xml

jreback commented on pandas/conftest.py:
```diff
@@ -437,6 +437,29 @@ def index(request):
+    # Azure pipeline that writes into indices_dict despite copy
+    ind = indices_dict[request.param].copy(deep=True)
+    vals = ind.values
+    if type(vals[0]) == tuple:
```
can you just check `request.parm` == 'multindex'? as its more explict

AlexKirko replied:
Done.

jreback commented on pandas/core/sorting.py:
```diff
@@ -306,10 +306,17 @@ def nargsort(
+    try:
+        indexer = non_nan_idx[non_nans.argsort(kind=kind)]
+    except TypeError:
+        # For compatibility with Series: fall back to quicksort
```
we have a test that hits this?

AlexKirko replied:
Sorry, this is a remnant from trying to bring `Series.sort_values` in line with `Index.sort_values`.  Reverted for now. This change doesn't belong in this PR. I'll bring it back in the next PR when I'll be synchronizing `Index` and `Series` behavior.

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +398,40 @@ def test_astype_preserves_name(self, index, dtype):
+def test_sort_values_invalid_na_position(index_with_missing, na_position):
+
+    if type(index_with_missing) in [DatetimeIndex, PeriodIndex, TimedeltaIndex]:
+        pytest.xfail("stable descending order sort not implemented")
```
e.g. maybe just enumerate the PeriodIndex case here (do the others have an issue?) pls open if they don't (you can make a single issue with checkboxes)

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +398,40 @@ def test_astype_preserves_name(self, index, dtype):
+    # GH 35584. Test that sort_values works with missing values,
+    # sort non-missing and place missing according to na_position
+
+    if type(index_with_missing) in [DatetimeIndex, PeriodIndex, TimedeltaIndex]:
```
same

AlexKirko replied:
@jreback 
Sorry, I wasn't at all clear in the xfail messages as to how the tests are different. The one in `period/test_ops.py` checks whether `PeriodIndex` and another non-datetime-like subtype sort duplicates into the same order when `descending=True`. As part of syncing duplicate sorting behavior in the next PR, I'd like to write a general test with an `index_with_duplicates` fixture to check for sorting stability and replace this test.

The tests in `test_common.py` xfail, because the `na_position` kwarg isn't implemented yet for datetime-like indices. Implementing it would go with implementing stable duplicate sorting behavior and break a bunch of tests. I also intend to deal with that in the next PR as I bring sort_values in sync between all Index subtypes and Series.

I've clarified the xfail messages everywhere and added comments with xrefs to the tests in `test_common.py`.  Is that acceptable? I think keeping the tests separate makes sense for now.

Sorry if I misunderstood your question.

jreback reviewed (changes requested):
looks good, couple more comments

AlexKirko committed:
TST: check for MultiIndex through request.param

AlexKirko committed:
CLN: revert changes to nargsort

AlexKirko committed:
DOC: add issue xref to xfail reason in period/test_ops.py

AlexKirko committed:
DOC: clarify xfail reasons and add comments

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +398,45 @@ def test_astype_preserves_name(self, index, dtype):
+        pytest.xfail("sort_values does not support na_position kwarg")
+    elif type(index_with_missing) in [CategoricalIndex, MultiIndex]:
+        pytest.xfail("missing value sorting order not defined for index type")
+    missing_count = np.sum(index_with_missing.isna())
```
can you add a blank line here

AlexKirko replied:
Done.

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -391,3 +398,45 @@ def test_astype_preserves_name(self, index, dtype):
+        # synchronizing duplicate-sorting behavior, because we currently expect
+        # them, other indices, and Series to sort differently (xref 35922)
+        pytest.xfail("sort_values does not support na_position kwarg")
+    elif type(index_with_missing) in [CategoricalIndex, MultiIndex]:
```
can you isinstance(....) instead

AlexKirko replied:
Done.

jreback reviewed (changes requested):
small comments, ping on green.

AlexKirko committed:
REFACT: switch to isinstance, add blank line

AlexKirko commented:
@jreback 
Changes made, all green.

jreback commented:
thanks @AlexKirko very nice! happy to take more to consolidate the Index/Series code for this :->

AlexKirko commented:
Thanks!
A bit sick atm, but I'll get back to this in a few days with a new PR.
